clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,232 @@
+    /// let a = String::new();
+    /// let closure = move || {
+    ///     drop(a);
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a = String::new();
+    /// let closure = || {
+    ///     drop(a);
+    /// };
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub NEEDLESS_MOVE,
+    restriction,
+    "checks for needless `move`s on closures / `async` blocks"
+}
+
+declare_lint_pass!(NeedlessMove => [NEEDLESS_MOVE]);
+
+impl NeedlessMove {
+    fn check_closure<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, closure: &'tcx Closure<'tcx>) {
+        let CaptureBy::Value { move_kw } = closure.capture_clause else {
+            return;
+        };
+
+        if move_kw.is_dummy() {
+            // async fn ...() {} convert the body to an `async move {}` block,
+            // with a DUMMY_SP for the move_kw
+            return;
+        }
+
+        // Collect moved & borrowed variables from the closure, which the closure *actually* needs.
+        let ctx = {
+            let mut ctx = MovedVariablesCtxt::default();
+            let body = cx.tcx.hir().body(closure.body);
+            let infcx = cx.tcx.infer_ctxt().build();
+            euv::ExprUseVisitor::new(&mut ctx, &infcx, closure.def_id, cx.param_env, cx.typeck_results())
+                .consume_body(body);
+            ctx
+        };
+
+        let mut lint_result = LintResult::NothingCaptured;
+
+        for captured_place in cx.typeck_results().closure_min_captures_flattened(closure.def_id) {
+            let place = &captured_place.place;
+            if let Some(ck_expr_id) = captured_place.info.capture_kind_expr_id {
+                let required_ck = ctx.get_required_kind(place, ck_expr_id);
+                match required_ck {
+                    UpvarCapture::ByRef(_) => {
+                        // no matter what the old `lint_result` is, we keep the move.
+                        lint_result = LintResult::NeedMove;
+                    },
+                    UpvarCapture::ByValue => {
+                        lint_result = match lint_result {
+                            LintResult::NothingCaptured | LintResult::Consumed => LintResult::Consumed,
+                            LintResult::NeedMove => LintResult::NeedMove,
+                        }
+                    },
+                }
+            }
+        }
+
+        let note_msg = match lint_result {
+            LintResult::NothingCaptured => "there were no captured variables, so the `move` is unnecessary",
+            LintResult::Consumed => {
+                "there were consumed variables, but no borrowed variables, so the `move` is unnecessary"
+            },
+            LintResult::NeedMove => {
+                // there was a value which would be borrowed if it weren't for the move keyword,
+                // so we should keep it, as removing it would change semantics.
+                return;
+            },
+        };
+
+        span_lint_and_then(
```

Comment:
We should call `is_from_proc_macro` here. You can implement `WithSearchPat` for `Closure` by reusing the code for `ExprKind::Closure` in `expr_search_pat`